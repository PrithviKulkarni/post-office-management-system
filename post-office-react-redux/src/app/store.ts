import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
