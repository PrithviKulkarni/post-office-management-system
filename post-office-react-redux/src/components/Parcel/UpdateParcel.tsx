import { useAppDispatch } from '../../app/hooks'
import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import { useSelector } from 'react-redux'
import { RootState } from '../../app/store'
import { getParcelById, updateAParcel } from '../../features/ParcelSlice'
import { useParams } from 'react-router-dom'
import { Parcel, defaultParcel } from '../interfaces/parcelInterface'
import { Container, Button, Form, InputGroup } from 'react-bootstrap'
import {  SyntheticEvent } from 'react'
/****
 * 
 * @returns shows an update form to edit the parcel details
 */
const UpdateParcel = () => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const params = useParams();
  const parcels = useSelector((state: RootState) => getParcelById(state, params.parcelId!))

  const [parcel_no, setParcelNo] = useState<number| undefined>(parcels?.parcel_no)
  const [parcel_weight, setParcelW] = useState<number| undefined>(parcels?.parcel_weight)
  const [status, setStatus] = useState<string| undefined>(parcels?.status)
  const [estimated_delivery_days, setEDD] = useState<number| undefined>(parcels?.estimated_delivery_days)
  const [errorMessage, setErrorMessage] = useState<string>('')



  const submitHandler = (event: SyntheticEvent) => {
    console.log('inside SUbmit Handler')
    event.preventDefault()
    navigate('/parcels')

    const validated = [
      parcel_no,
      parcel_weight,
      status,
      estimated_delivery_days
    ].every(Boolean)

    try {
      if (validated) {
        const updParcelData: Parcel = defaultParcel
        updParcelData.id = params.parcelId!

        if (parcel_no) {updParcelData.parcel_no = parcel_no}
        if (parcel_weight) {
          updParcelData.parcel_weight = parcel_weight
        }
        if (status) {
          updParcelData.status = status
        }
        if (estimated_delivery_days) {
          updParcelData.estimated_delivery_days = 
            estimated_delivery_days
          
        }
         console.log({updParcelData});
        dispatch(updateAParcel(updParcelData))
        navigate('/parcels')
      }
    } catch (error) {
      setErrorMessage('Parcel NOT FOUND.')
      console.log('errorMessage: ', errorMessage)
    }
  }


    return (
      <Container>
        <h3 className='text-center text-light'> Edit the parcel </h3>

        <Form onSubmit={submitHandler} className="updateParcel" data-testid="updateParcelTest" >
          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>Parcel no</InputGroup.Text>
            <Form.Control
              type='text'
              data-testid = 'parcelNoInput'
              aria-describedby='basic-addon1'
              size='sm'
              name='parcel_no'
              value={parcel_no}
              onChange={event => setParcelNo(Number(event.target.value))}
            />
          </InputGroup>

          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>
              Parcel Weight(kg)
            </InputGroup.Text>
            <Form.Control
              type='text'
              data-testid = "weightInput"
              aria-describedby='basic-addon1'
              size='sm'
              name='parcel_weight'
              value={parcel_weight}
              onChange={event => setParcelW(Number(event.target.value))}
            />
          </InputGroup>

          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>Status</InputGroup.Text>
            <Form.Control
              type='text'
              data-testid="statusInput"
              aria-describedby='basic-addon1'
              size='sm'
              name='status'
              value={status}
              onChange={event => setStatus(event.target.value)}
            />
          </InputGroup>

          <InputGroup className='mb-3'>
            <InputGroup.Text id='basic-addon1'>
              Estimated delivery (in days)
            </InputGroup.Text>
            <Form.Control
              type='text'
              data-testid="eddInput"
              aria-describedby='basic-addon1'
              size='sm'
              name='estimated_delivery_days'
              value={estimated_delivery_days}
              onChange={event => setEDD(Number(event.target.value))}
            />
          </InputGroup>

          <Button type='submit' data-testid="updateButton">Edit Parcel</Button>
          <p>{errorMessage}</p>
        </Form>
      </Container>
    )
  }


export default UpdateParcel
