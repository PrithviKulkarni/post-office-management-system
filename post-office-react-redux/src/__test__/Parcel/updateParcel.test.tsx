import React from "react";
import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import { store } from "../../app/store";
import { Provider } from "react-redux";
import UpdateParcel from "../../components/Parcel/UpdateParcel";
import { BrowserRouter } from "react-router-dom";

afterEach(cleanup);
it("renders Update Parcel page correctly", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BrowserRouter>
        <UpdateParcel />
      </BrowserRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("Verifies the Update Parcel render", () => {
  const{getByTestId} = render(
    <Provider store={store}>
      <BrowserRouter>
        <UpdateParcel />
      </BrowserRouter>
    </Provider>
  );


  const button = getByTestId("updateButton");
  expect(button.textContent).toBe("Edit Parcel");
  

  const parcelNo = getByTestId("parcelNoInput")
    const parcelWeight = getByTestId("weightInput")
    const status = getByTestId("statusInput")
    const edd = getByTestId("eddInput")
    fireEvent.change(parcelNo, {target: {value: "1"}});
    fireEvent.change(parcelWeight, {target: {value: "2"}})
    fireEvent.change(status, {target: {value: "test"}})
    fireEvent.change(edd, {target: {value: "3"}} )
    fireEvent.click(button);
    expect(button).toBeTruthy();
  
  expect(screen.getByTestId("updateParcelTest")).toHaveClass("updateParcel");
});
