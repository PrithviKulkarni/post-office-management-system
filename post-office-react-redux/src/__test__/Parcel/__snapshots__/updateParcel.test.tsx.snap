// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Update Parcel page correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="text-center text-light"
    >
       Edit the parcel 
    </h3>
    <form
      class="updateParcel"
      data-testid="updateParcelTest"
    >
      <div
        class="mb-3 input-group"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Parcel no
        </span>
        <input
          aria-describedby="basic-addon1"
          class="form-control form-control-sm"
          data-testid="parcelNoInput"
          name="parcel_no"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3 input-group"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Parcel Weight(kg)
        </span>
        <input
          aria-describedby="basic-addon1"
          class="form-control form-control-sm"
          data-testid="weightInput"
          name="parcel_weight"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3 input-group"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Status
        </span>
        <input
          aria-describedby="basic-addon1"
          class="form-control form-control-sm"
          data-testid="statusInput"
          name="status"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3 input-group"
      >
        <span
          class="input-group-text"
          id="basic-addon1"
        >
          Estimated delivery (in days)
        </span>
        <input
          aria-describedby="basic-addon1"
          class="form-control form-control-sm"
          data-testid="eddInput"
          name="estimated_delivery_days"
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        data-testid="updateButton"
        type="submit"
      >
        Edit Parcel
      </button>
      <p />
    </form>
  </div>
</DocumentFragment>
`;
